tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc

```diff
@@ -120,6 +120,135 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
 
+namespace {
+bool Is1dSliceWithoutStrides(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kSlice &&
+         1 == instr->slice_starts().size() &&
+         1 == instr->slice_limits().size() &&
+         1 == instr->slice_strides().size() &&
+         1 == instr->slice_strides().at(0);
+}
+
+bool IsSliceInputFusion(const HloInstruction& unnested_hlo) {
+  if (!unnested_hlo.IsInputFusion()) {
+    return false;
+  }
+
+  const HloInstruction* root = unnested_hlo.fused_expression_root();
+  if (Is1dSliceWithoutStrides(root)) {
+    return true;
+  }
+
+  if (root->opcode() != HloOpcode::kTuple) {
+    return false;
+  }
+
+  return absl::c_all_of(root->operands(), [](const HloInstruction* instr) {
+    return Is1dSliceWithoutStrides(instr);
+  });
+}
+
+bool ConcatHasNoEffect(const HloInstruction* concat) {
+  // Check if this concat is in the below pattern. In addition, we check
+  // that the slices combiningly are in effect a reverse function of the
+  // concat.
+  //
+  //     Concat
+  //     |    |
+  //     v    v
+  //   Slice Slice
+  //
+  std::vector<HloInstruction*> users = concat->users();
+  bool all_1d_slices = absl::c_all_of(users, [](const HloInstruction* i) {
+    return Is1dSliceWithoutStrides(i);
+  });
+  if (!all_1d_slices) {
+    // Limit our supported cases to 1 dimensional slices.
+    return false;
+  }
+  absl::c_sort(users, [](const HloInstruction* a, const HloInstruction* b) {
+    return a->slice_starts().at(0) < b->slice_starts().at(0);
+  });
+
+  // Verify that each operand to the concat is reversed by a slice.
+  if (users.size() != concat->operand_count()) {
+    return false;
+  }
+  int64 prev_limit = 0;
+  for (size_t i = 0; i < users.size(); ++i) {
+    const HloInstruction* u = users[i];
+    int64 slice_size = u->slice_limits().at(0) - u->slice_starts().at(0);
+    if (u->slice_starts().at(0) != prev_limit ||
+        slice_size != ShapeUtil::ElementsIn(concat->operand(i)->shape())) {
+      return false;
+    }
+    prev_limit = u->slice_limits().at(0);
+  }
+
+  return true;
+}
+
+// Returns whether we can prove the transitive uses are in effect elementwise
+// operations. A concat followed by slices is considered (effectively)
+// elementwise if the slices combiningly is a reverse function of the concat.
+// We can prove more patterns but we currently do just enough for
+// SliceInputFusion.
+bool AreTransitiveUsesEffectivelyElementwise(const HloInstruction* instr) {
+  absl::flat_hash_set<const HloInstruction*> visited;
+  absl::InlinedVector<const HloInstruction*, 4> stack;
+  stack.push_back(instr);
+  while (!stack.empty()) {
+    const HloInstruction* current = stack.back();
+    stack.pop_back();
+    visited.insert(current);
+    for (const HloInstruction* user : current->users()) {
+      VLOG(3) << "Visiting: " << user->ToString();
+      switch (user->opcode()) {
+        case HloOpcode::kTuple:
+        // We say reshape is fine because it does not reorder elements.
```

Comment:
There is the ReshapeIsBitcast function which you could use here to verify this.
Theoretically Reshape could reorder elements, but we usually try to assign a layout so that it doesn't.

Reply:
Since we check row-major layout below, it is safe to assume reshape does not re-order elements. However, ShapeBitcast is better. Thanks for the pointer.